cpp/src/arrow/array/statistics.h

```diff
@@ -0,0 +1,75 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <variant>
+
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+/// \brief Statistics for an Array
+///
+/// Apache Arrow format doesn't have statistics but data source such
+/// as Apache Parquet may have statistics. Statistics associated with
+/// data source can be read unified API via this class.
+struct ARROW_EXPORT ArrayStatistics {
+  using ValueType =
+      std::variant<bool, int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t, int64_t,
+                   uint64_t, std::string, std::string_view>;
+
+  ArrayStatistics() = default;
+  ~ArrayStatistics() = default;
+
+  /// \brief The number of null values, may not be set
+  std::optional<int64_t> null_count = std::nullopt;
+
+  /// \brief The number of distinct values, may not be set
+  std::optional<int64_t> distinct_count = std::nullopt;
+
+  /// \brief The minimum value, may not be set
+  std::optional<ValueType> min = std::nullopt;
+
+  /// \brief Whether the minimum value is exact or not, may not be set
+  std::optional<bool> is_min_exact = std::nullopt;
```

Comment:
If not set, it means "might not exact"?

Reply:
Yes. It means "unknown".

Reply:
Well, if it's unknown, then it's not exact. The third state is not useful IMHO.

Reply:
Hmm. It may be exact.

`is_max_value_exact` in the Parquet's definition is optional:
https://github.com/apache/arrow/blob/d078d5cafc53823f23bdab9918611b65a54f21be/cpp/src/parquet/parquet.thrift#L279-L282

We may need `std::optional` to support Parquet statistics. Can we keep this `std::optional` for now?

Reply:
I think the reason it's optional in Parquet is simply for compatibility with older files. @wgtmac Is it right?

Reply:
> Well, if it's unknown, then it's not exact. The third state is not useful IMHO.

I agree with @pitrou. We can simply mark it as not exact if this field is missing.

Reply:
OK.
Let's remove `std::optional`: #43594